### DISCUSSION TITLE
Are optimistic locking or pessimistic locking used when updating wallet balances?

### DISCUSSION THREAD

**Discussion by 704403314** [bavix/laravel-wallet#814] [Q&A] [answered]
To ensure the accuracy of wallet balances under high concurrency, are optimistic locking or pessimistic locking used when updating wallet balances?
For example, if a user's current wallet balance is 0 and they initiate two refund transactions simultaneously, each refund requiring an increase of 100 units to the wallet balance. In a high concurrency situation, it is possible that both refund transactions start with a balance query of 0. Consequently, each transaction increases the balance by 100 units based on the initial balance of 0, resulting in a final wallet balance of 100 units for the user, whereas it should have been 200 units.

**Comment by rez1dent3** [comment, marked as answer]
Hello. It depends on the cache and lock settings. If you leave array, then there will be no locks.

I recommend using redis sentinel for cache and lock. But you can use a database for cache, then you will get full-fledged transactions at the database level.

https://github.com/bavix/laravel-wallet/blob/9b0bd38bd07d3d251cb6a48ae2eb11269a82342b/config/config.php#L65-L79

Hm. It will be necessary to add a functional test for the database as cache... Thank you)
